### PR TITLE
feat(tile-view): initial implementation for mobile

### PR DESCRIPTION
**Pull request by virtuacoplenny**
- Create a tile view component for displaying thumbnails in a
  two-dimensional grid.
- Update the existing TileViewButton so it shows a label in the
  overflow menu.
- Modify conference so it can display TileView while hiding
  Filmstrip.
- Modify Thumbnail so its width/height can be set and to prevent
  pinning while in tile view mode.

### CODE REVIEW

File: react/features/filmstrip/components/native/Thumbnail.js [hunk 1]
```diff
@@ -57,19 +58,27 @@ class Thumbnail extends Component {
...
-        if (participant.pinned) {
+        if (participant.pinned && !disablePin) {
             style = {
                 ...style,
                 ...styles.thumbnailPinned
             };
         }
 
+        if (this.props.styleOverrides) {
+            style = {
```
Comment (saghul) [inline react/features/filmstrip/components/native/Thumbnail.js]: why not use style arrays? That way it doesn't matter if styleOverrides is undefined.

File: react/features/filmstrip/components/native/Thumbnail.js [hunk 2]
```diff
@@ -96,7 +105,7 @@ class Thumbnail extends Component {
                 <ParticipantView
                     avatarSize = { AVATAR_SIZE }
                     participantId = { participantId }
-                    tintEnabled = { participantInLargeVideo }
+                    tintEnabled = { participantInLargeVideo && !disablePin }
```
Comment (saghul) [inline react/features/filmstrip/components/native/Thumbnail.js]: Why is this needed? `participantInLargeVideo` will always be false in tile view anyway. Am I missing something?

File: react/features/filmstrip/components/native/TileView.js [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        if (this.props._isNarrowAspectRatio) {
+            return participantCount >= 3 ? 2 : 1;
+        }
+
+        if (participantCount === 4) {
+            // In wide view, a four person call should display as a 2x2 grid.
+            return 2;
+        }
+
+        return participantCount >= 3 ? 3 : participantCount;
```
Comment (saghul) [inline react/features/filmstrip/components/native/TileView.js]: `Math.min(3, participantCount)` ?

File: react/features/filmstrip/components/native/TileView.js [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    /**
+     * Returns the current height and width of the screen.
+     *
+     * @private
+     * @returns {Object}
+     */
+    _getScreenDimensions() {
```
Comment (saghul) [inline react/features/filmstrip/components/native/TileView.js]: This won't work when multiple apps are used at once. Can you please use this to measure the available surface? https://github.com/jitsi/jitsi-meet/blob/master/react/features/base/responsive-ui/components/DimensionsDetector.native.js

File: react/features/filmstrip/components/native/Thumbnail.js (hunk 2 above)
Comment (virtuacoplenny) [reply react/features/filmstrip/components/native/Thumbnail.js]: Nothing clears large video state when switching to tile view. Large video updates are allowed to proceed on web because electron always-on-top needs to know who is on large video. I could try to find a way to clear the state, but maybe on mobile only.

Comment (saghul) [reply react/features/filmstrip/components/native/Thumbnail.js]: > Edit: at the very, very least, I probably shouldn't re-use the disablePin prop to prevent tinting.

👍

**Comment by virtuacoplenny** [comment]
Force pushed to resolve conflicts in react/features/video-layout/functions.js.

File: react/features/filmstrip/components/native/Thumbnail.js [hunk 5]
```diff
@@ -18,31 +19,32 @@ import { AVATAR_SIZE } from '../styles';
 import styles from './styles';
 import VideoMutedIndicator from './VideoMutedIndicator';
 
+/**
+ * Thumbnail component's property types.
+ */
+type Props = {
```
Comment (zbettenbuk) [inline react/features/filmstrip/components/native/Thumbnail.js]: There were a lot of already existing props here, but can you please add docs to those that you added?

File: react/features/filmstrip/components/native/TileView.js [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+    getNearestReceiverVideoQualityLevel,
+    setMaxReceiverVideoQuality
+} from '../../../base/conference';
+import {
+    DimensionsDetector,
+    isNarrowAspectRatio,
+    makeAspectRatioAware
+} from '../../../base/responsive-ui';
+
+import Thumbnail from './Thumbnail';
```
Comment (zbettenbuk) [inline react/features/filmstrip/components/native/TileView.js]: Please put in alphabetical order

File: react/features/filmstrip/components/native/TileView.js [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+
+/**
+ * The type of the React {@link Component} props of {@link TileView}.
+ */
+type Props = {
+
+    /**
+     * The participants in the conference.
+     */
+    _participants: Array<any>,
```
Comment (zbettenbuk) [inline react/features/filmstrip/components/native/TileView.js]: Shouldn't it be Array &lt; Object &gt; ?

File: react/features/filmstrip/components/native/TileView.js [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    /**
+     * The available height for {@link TileView} to occupy.
+     */
+    width: number
+};
+
+const TILE_ASPECT_RATIO = 1;
+
+/**
+ * Implements a React {@link Component} which represents the filmstrip on
```
Comment (zbettenbuk) [inline react/features/filmstrip/components/native/TileView.js]: Copy paste ;)

File: react/features/filmstrip/components/native/TileView.js [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+     */
+    constructor(props: Props) {
+        super(props);
+
+        // Bind event handler so it is only bound once per instance.
+        this._onDimensionsChanged = this._onDimensionsChanged.bind(this);
+    }
+
+    /**
+     * Updates the receiver video quality.
```
Comment (zbettenbuk) [inline react/features/filmstrip/components/native/TileView.js]: It does that for sure, but wouldn't it be more flexible to use a doc related to ```componentDidMount``` so then we don't need to change it if we add more logic? (same below)

File: react/features/filmstrip/components/native/TileView.js [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+                    style = {{
+                        ...styles.tileView,
+                        height,
+                        width
+                    }}>
+                    <TouchableWithoutFeedback onPress = { onClick }>
+                        <View
+                            style = {{
+                                ...styles.tileViewRows,
+                                minHeight: height,
```
Comment (zbettenbuk) [inline react/features/filmstrip/components/native/TileView.js]: Do these style props do anything here?

File: react/features/filmstrip/components/native/TileView.js (hunk 6 above)
Comment (virtuacoplenny) [reply react/features/filmstrip/components/native/TileView.js]: You're talking about `./Thumbnail` and `./styles`? I think this is the desired order. At some point it was decided capitals should come before lowercase.

File: react/features/filmstrip/components/native/TileView.js (hunk 10 above)
Comment (virtuacoplenny) [reply react/features/filmstrip/components/native/TileView.js]: Yes they do stuff; I tested them again just now to make sure. `style.tileViewRows` is used to make sure the tiles are vertical aligned center on the screen. minHeight and minWidth are used to make sure the container takes up at least the full screen; that way when there are only a few thumbnails, the thumbnails will display in the middle of the screen thanks to center alignment. minX is used so that the container is allowed to grow as needed, beyond the size of the parent container, so that thumbnails can become scrollable. Arguably minWidth isn't really needed because there should be no horizontal scrolling. So that's my answer and I don't claim it to be the right way but was the culmination of me playing around for a while.

File: react/features/base/react/components/AbstractContainer.js [hunk 11]
```diff
@@ -31,7 +31,10 @@ export default class AbstractContainer extends Component<*> {
          * The style (as in stylesheet) to be applied to this
          * {@code AbstractContainer}.
          */
-        style: PropTypes.object,
+        style: PropTypes.oneOfType([
```
Comment (saghul) [inline react/features/base/react/components/AbstractContainer.js]: @virtuacoplenny Since you are here, can you change this to Flow types? There is an unwritten rule (that I just made up) which says that if you touch `PropTypes` code you automatically volunteer to change it to Flow types.

File: react/features/conference/components/Conference.native.js [hunk 12]
```diff
@@ -245,20 +259,23 @@ class Conference extends Component<Props> {
 
                 {/*
                   * The LargeVideo is the lowermost stacking layer.
-                  */}
-                <LargeVideo onPress = { this._onClick } />
+                  */
+                    _shouldDisplayTileView
+                        ? <TileView onClick = { this._onClick } />
+                        : <LargeVideo onPress = { this._onClick } />
```
Comment (saghul) [inline react/features/conference/components/Conference.native.js]: I know it's not your code, but can you please change that onPress to onClick?

File: react/features/filmstrip/components/native/TileView.js [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+     * Returns all participants with the local participant at the end.
+     *
+     * @private
+     * @returns {Participant[]}
+     */
+    _getSortedParticipants() {
+        const participants = [];
+        let localParticipant;
+
+        this.props._participants.forEach(participant => {
```
Comment (saghul) [inline react/features/filmstrip/components/native/TileView.js]: for..of

File: react/features/filmstrip/components/native/TileView.js [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+            if (participant.local) {
+                localParticipant = participant;
+            } else {
+                participants.push(participant);
+            }
+        });
+
+        if (localParticipant) {
```
Comment (saghul) [inline react/features/filmstrip/components/native/TileView.js]: use &&

File: react/features/filmstrip/components/native/TileView.js [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+     *
+     * @private
+     * @returns {Object}
+     */
+    _getTileDimensions() {
+        const { _participants } = this.props;
+        const { height, width } = this.state;
+        const columns = this._getColumnCount();
+        const participantCount = _participants.length;
+        const heightToUse = height - 20;
```
Comment (saghul) [inline react/features/filmstrip/components/native/TileView.js]: where does this 20 come from? maybe make it a constant?

File: react/features/toolbox/components/native/OverflowMenu.js [hunk 16]
```diff
@@ -72,6 +73,7 @@ class OverflowMenu extends Component<Props> {
                 <RecordButton { ...buttonProps } />
                 <LiveStreamButton { ...buttonProps } />
                 <PictureInPictureButton { ...buttonProps } />
+                <TileViewButton { ...buttonProps } />
```
Comment (saghul) [inline react/features/toolbox/components/native/OverflowMenu.js]: Can you please put this as the previous to last? We have kept the PiP button as the last option for a while and IMHO it's nice to have it that way.

File: react/features/video-layout/functions.js [hunk 17]
```diff
@@ -73,7 +73,8 @@ export function shouldDisplayTileView(state: Object = {}) {
     return Boolean(
         state['features/video-layout']
             && state['features/video-layout'].tileViewEnabled
-            && !state['features/etherpad'].editing
+            && (!state['features/etherpad']
```
Comment (saghul) [inline react/features/video-layout/functions.js]: I can't wait until we update RN to 0.56, which uses Babel 7 so we can use optional chaining: https://babeljs.io/docs/en/next/babel-plugin-proposal-optional-chaining.html

File: react/features/base/react/components/AbstractContainer.js (hunk 11 above)
Comment (virtuacoplenny) [reply react/features/base/react/components/AbstractContainer.js]: I tried and the subsequent flow errors from other consumers of AbstractContainer were too much for me to handle.

File: react/features/filmstrip/components/native/TileView.js (hunk 15 above)
Comment (virtuacoplenny) [reply react/features/filmstrip/components/native/TileView.js]: This is taking away 20 (10 from each side) of the tile view container so that the tiles aren't up against the edge. I've moved the 10 to a constant.

**Comment by virtuacoplenny** [comment]
Force pushed to fix conflicts in the import statements within Conference.native.js and OverflowMenu.js.

File: react/features/filmstrip/components/native/TileView.js (hunk 6 above)
Comment (zbettenbuk) [reply react/features/filmstrip/components/native/TileView.js]: Grr... I hate these unwritten rules. My understanding is that capitals come first only if they are the same letter. E.g. 'T' comes before 't', but won't come before 's'. But I may be wrong, so just ignore it, I'm fine this way too

File: react/features/filmstrip/components/native/TileView.js (hunk 10 above)
Comment (zbettenbuk) [reply react/features/filmstrip/components/native/TileView.js]: Hm, interesting, I didn't know it would work like this. Okay, thanks!

File: react/features/base/react/components/AbstractContainer.js (hunk 11 above)
Comment (zbettenbuk) [reply react/features/base/react/components/AbstractContainer.js]: I'm happy to help, if need be. I love battling with flow :)

Comment (virtuacoplenny) [reply react/features/base/react/components/AbstractContainer.js]: I did the stupid thing of writing code until it works. Passes flow now.

**Comment by virtuacoplenny** [comment]
Force pushed to resolve conflicts in Conference.native due to the DialogContainer being moved.

**Comment by virtuacoplenny** [comment]
@saghul, can you do another pass?
